compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +152,208 @@ pub fn expand_deriving_smart_ptr(
+                })
+                .cloned()
+                .collect();
+            let mut substitution = TypeSubstitution {
+                from_name: pointee_ty_ident.name,
+                to_ty: &s_ty,
+                rewritten: bounds.len() != bound.bounds.len(),
+            };
+            let mut predicate = ast::WherePredicate::BoundPredicate(ast::WhereBoundPredicate {
+                span: bound.span,
+                bound_generic_params: bound.bound_generic_params.clone(),
+                bounded_ty: bound.bounded_ty.clone(),
+                bounds,
+            });
+            substitution.visit_where_predicate(&mut predicate);
+            if substitution.rewritten {
+                impl_generics.where_clause.predicates.push(predicate);
+            }
+        }
+    }
+
+    let extra_param = cx.typaram(span, Ident::new(sym::__S, span), self_bounds, None);
+    impl_generics.params.insert(pointee_param_idx + 1, extra_param);
 
     // Add the impl blocks for `DispatchFromDyn` and `CoerceUnsized`.
     let gen_args = vec![GenericArg::Type(alt_self_type.clone())];
     add_impl_block(impl_generics.clone(), sym::DispatchFromDyn, gen_args.clone());
     add_impl_block(impl_generics.clone(), sym::CoerceUnsized, gen_args.clone());
 }
+
+fn is_sized_marker(path: &ast::Path) -> bool {
+    const CORE_UNSIZE: [Symbol; 3] = symbols!(core::marker::Sized);
+    const STD_UNSIZE: [Symbol; 3] = symbols!(std::marker::Sized);
+    if path.segments.len() == 3 {
+        path.segments.iter().zip(CORE_UNSIZE).all(|(segment, symbol)| segment.ident.name == symbol)
+            || path
+                .segments
+                .iter()
+                .zip(STD_UNSIZE)
+                .all(|(segment, symbol)| segment.ident.name == symbol)
+    } else {
+        *path == sym::Sized
+    }
+}
+
+struct TypeSubstitution<'a> {
+    from_name: Symbol,
+    to_ty: &'a AstTy,
+    rewritten: bool,
+}
+
+impl<'a> ast::mut_visit::MutVisitor for TypeSubstitution<'a> {
+    fn visit_ty(&mut self, ty: &mut AstTy) {
+        if let Some(name) = ty.kind.is_simple_path()
+            && name == self.from_name
+        {
+            *ty = self.to_ty.clone();
+            self.rewritten = true;
+            return;
+        }
+        match &mut ty.kind {
+            ast::TyKind::Slice(_)
+            | ast::TyKind::Array(_, _)
+            | ast::TyKind::Ptr(_)
+            | ast::TyKind::Ref(_, _)
+            | ast::TyKind::BareFn(_)
+            | ast::TyKind::Never
+            | ast::TyKind::Tup(_)
+            | ast::TyKind::AnonStruct(_, _)
+            | ast::TyKind::AnonUnion(_, _)
+            | ast::TyKind::Path(_, _)
+            | ast::TyKind::TraitObject(_, _)
+            | ast::TyKind::ImplTrait(_, _)
+            | ast::TyKind::Paren(_)
+            | ast::TyKind::Typeof(_)
+            | ast::TyKind::Infer
+            | ast::TyKind::MacCall(_)
+            | ast::TyKind::Pat(_, _) => ast::mut_visit::walk_ty(self, ty),
```

Comment:
I don't think it makes sense to exhaustively match here -- you can just replace the whole match block with a `walk_ty` call.

Reply:
Right. Replaced.